funnel/models/project.py

```diff
@@ -879,6 +879,14 @@ def roles_for(self, actor=None, anchors=()):
             if crew_membership is not None:
                 roles.update(crew_membership.offered_roles())
 
+            if (
+                self.rsvp_for(actor) is not None
+                or self.participants.filter_by(user=actor).one_or_none() is not None
+            ):
+                roles.add('attendee')
+
+            if 'attendee' in roles or 'crew' in roles:
+                roles.add('participant')
```

Comment:
This will continue to work using the grants mechanism as "roles" here is a LazyRoleSet. It will look for existing role grants.